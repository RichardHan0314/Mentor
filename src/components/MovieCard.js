/* eslint-disable react/forbid-prop-types */
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import MovieAction from '../containers/MovieAction';

const MovieCard = (props) => {
    const { movie, title, genres, date, image } = props;

    const [isHovered, setHovered] = useState(false);

    const handleHover = () => {
        setHovered(!isHovered);
    };

    return (
        <Wrapper 
            onMouseEnter={handleHover}
            onMouseLeave={handleHover}>
            <Poster image={image}>
                <MovieAction 
                    {...{isHovered}}
                    movie = {movie}
                />
            </Poster>
            <MovieInfo>
                <MovieInfoCol>
                    <Title>{title}</Title>
                    <Genre>{genres.join(' ')}</Genre>
                </MovieInfoCol>
                <ReleaseDate>{date}</ReleaseDate>
            </MovieInfo>
        </Wrapper>
    );
};

const Wrapper = styled.section`
    width: 100%;
    margin: 0 0 20px 0;
`;

const Poster = styled.div`
    display: flex;
    justify-content: flex-end;
    width: 100%;
    height: 600px;
    background-image: url(${props => props.image});
    background-repeat: no-repeat;
    background-size: cover;
`;

const MovieInfo = styled.div`
    display: flex;
    justify-content: space-between;
    color: #fff;
    padding: 16px 0;
`;

const MovieInfoCol = styled.div`
    display: flex;
    flex-direction: column;
`;

const Title = styled.h3`
  margin: 0;
  padding: 4px 0;
`;

const Genre = styled.p`
    margin: 0;
    padding: 4px 0;
`;

const ReleaseDate = styled.p`
    display: table;
    margin: 0;
    padding: 5px 20px;
    border: 1px solid #fff;
    border-radius: 4px;
`;

MovieCard.propTypes = {
    image: PropTypes.string,
    title: PropTypes.string.isRequired,
    genres: PropTypes.array,
    date: PropTypes.string.isRequired,
    movie: PropTypes.object.isRequired
};

MovieCard.defaultProps = {
    image: 'https://cdn3.vectorstock.com/i/1000x1000/50/07/http-404-not-found-error-message-hypertext-vector-20025007.jpg',
    genres: []
};

export default MovieCard;
