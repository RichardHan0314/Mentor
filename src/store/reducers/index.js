/* eslint-disable import/prefer-default-export */
import { combineReducers } from 'redux';
import { moviesReducer } from './moviesReducer';
import { searchReducer } from './searchReducer';
import { filtersReducer } from './filtersReducer';
import { sortingReducer } from './sortingReducer';

export const rootReducer = combineReducers({
    movies: moviesReducer,
    search: searchReducer,
    filters: filtersReducer,
    sorting: sortingReducer
});
