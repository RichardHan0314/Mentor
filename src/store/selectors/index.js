/* eslint-disable import/prefer-default-export */
import { createSelector } from 'reselect';

export const getAllMovies = state => state.movies.movies;
export const searchValueSelector = state => state.search.searchValue;
export const getFilters = state => state.filters.filters;
export const getFiltersSelector = state => state.filters;
export const sortValueSelector = state => state.sorting.sortingValue;

// export const getFilter = (state) => {
//     // const { allFilters } = state.filters;
//     // const filterData = [];

//     // allFilters.filter(item => {
//     //     if (item.status === true) {
//     //         filterData.push(item.name);
//     //     }
//     //     return filterData;
//     // });

//     // return filterData;

//     const filters = getAllFilters(state);

//     return filters.reduce((filters, { status, name }) => {
//         return status ? [...filters, name] : filters;
//     }, [])
// };

// https://react-redux.js.org/api/hooks#equality-comparisons-and-updates
export const getActiveFilters = createSelector(getFilters, filters => {
    return Object.keys(filters).filter(key => filters[key]);
})

export const getParams = createSelector(
    [searchValueSelector, sortValueSelector, getActiveFilters],
    (search, sortBy, filter) => {
        return { search, searchBy: 'title', sortBy, filter: filter.join() }
    }
);