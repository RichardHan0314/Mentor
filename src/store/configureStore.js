/* eslint-disable import/prefer-default-export */
import { createStore, applyMiddleware, compose } from 'redux';
import logger from 'redux-logger';
import thunk from 'redux-thunk';
import { rootReducer} from '../reducers';
// import { ping } from './enhancers/ping'; // <-- подключаем наш enhancer

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(rootReducer, composeEnhancers(applyMiddleware(thunk, logger)));