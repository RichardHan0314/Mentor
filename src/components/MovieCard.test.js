import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import { MemoryRouter, Link } from 'react-router-dom';
import MovieCard from './MovieCard';
import { store } from '../store/configureStore';

describe('MovieCard component', () => {
    const moviesDataMock = {
        id: 1,
        image: 'https://images-na.ssl-images-amazon.com/images/I/71niXI3lxlL._AC_SY679_.jpg',
        title: 'Avengers',
        genres: ['Action', 'Cartoon'],
        release_date: '2016'
    };

    it('render MovieCard snapshot', () => {
        const component = shallow(
            <Provider store={store}>
                <MemoryRouter>
                    <MovieCard movie = {moviesDataMock}
                        key = {moviesDataMock.id}
                        image = {moviesDataMock.poster_path}
                        title = {moviesDataMock.title}
                        genres = {moviesDataMock.genres}
                        date = {moviesDataMock.release_date}/>
                </MemoryRouter>
            </Provider>
        );
        expect(component).toMatchSnapshot();
    });
});