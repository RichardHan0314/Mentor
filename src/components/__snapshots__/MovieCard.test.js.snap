// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard component render MovieCard snapshot 1`] = `ShallowWrapper {}`;
